compiler/rustc_passes/src/reachable.rs

```diff
@@ -205,22 +205,32 @@ impl<'tcx> ReachableContext<'tcx> {
                     }
 
                     hir::ItemKind::Const(_, _, _, init) => {
-                        // Only things actually ending up in the final constant value are reachable
-                        // for codegen. Everything else is only needed during const-eval, so even if
-                        // const-eval happens in a downstream crate, all they need is
-                        // `mir_for_ctfe`.
-                        match self.tcx.const_eval_poly_to_alloc(item.owner_id.def_id.into()) {
-                            Ok(alloc) => {
-                                let alloc = self.tcx.global_alloc(alloc.alloc_id).unwrap_memory();
-                                self.propagate_from_alloc(alloc);
+                        let predicates = self.tcx.predicates_of(item.owner_id);
+
+                        if predicates.predicates.is_empty()
+                            || !rustc_trait_selection::traits::impossible_predicates(
+                                self.tcx,
+                                predicates.instantiate_identity(self.tcx).predicates,
+                            )
+                        {
+                            // Only things actually ending up in the final constant value are reachable
+                            // for codegen. Everything else is only needed during const-eval, so even if
+                            // const-eval happens in a downstream crate, all they need is
+                            // `mir_for_ctfe`.
+                            match self.tcx.const_eval_poly_to_alloc(item.owner_id.def_id.into()) {
```

Comment:
This being one such place where we evaluate generic stuff where we probably oughtn't